src/Elastic.Apm/Metrics/IMetricsProvider.cs

```diff
@@ -0,0 +1,32 @@
+using System.Collections.Generic;
+using Elastic.Apm.Api;
+
+namespace Elastic.Apm.Metrics
+{
+	/// <summary>
+	/// Defines an interface that every class implements that provides some metric value.
+	/// This interface is known to the <see cref="MetricsCollector" /> type and you
+	/// can implement new providers for other metrics by implementing this interface
+	/// and adding it to <see cref="MetricsCollector" />.
+	/// </summary>
+	internal interface IMetricsProvider
+	{
+		/// <summary>
+		/// Stores the number of calls to the <see cref="GetValue"/> method when it either returned null or an empty list.
+		/// This is used by <see cref="MetricsCollector"/>
+		/// </summary>
+		int ConsecutiveNumberOfFailedReads { get; set; }
```

Comment:
I don’t agree with this one, it’s not `MetricsCollector`detail. Each `IMetricsProvider.cs` knowns it’s number of fails, yes currently it’s only used in `MetricsCollector`, but I don’t see any reason having an extra collection in `MetricsCollector` to know the number of consecutive failures for each provider - I think that information belongs to the provider.

Reply:
The reason I think this is important is because we want to make `IMetricsProvider` part of public API to allow users to add metrics. Currently `IMetricsProvider` violates [Interface segregation principle](https://en.wikipedia.org/wiki/Interface_segregation_principle) - it should handle getting the metrics and not error handling/recovery as well. What if for the next release we decide to tweak error handling/recovery - for example to stop a metric provider based on `TotalNumberOfFailedReads` in addition/or instead of `ConsecutiveNumberOfFailedReads`. It will needlessly break backward compatibility with existing `IMetricsProvider` implementations even though none of them had much to do with error handling/recovery to begin with - we just used them as "bags of fields" to keep data.

Regarding the implementation - it doesn't have to be a separate data structure, for example we can store (IMetricsProvider, counter) pairs in what currently is a list of `IMetricsProvider`-s.

Reply:
Also if `IMetricsProvider` is to be responsible for error handling/recovery as well then it should encapsulated in it according to OOP best practices and not just be data carrier for `MetricsCollector`.

Reply:
As we discussed if we don't plan to expose this interface as part of public API we can leave as it's.